File: DeliveryBot/DeliveryBot.Db/Models/Robot.cs
﻿using System.ComponentModel.DataAnnotations.Schema;
using Point = NetTopologySuite.Geometries.Point;

namespace DeliveryBot.Db.Models;

public class Robot : Entity
{
    public string? Name { get; set; }
    public RobotStatus Status { get; set; }
    public int BatteryCharge { get; set; }
    public decimal WeightCapacityG { get; set; }
    public decimal VolumeCapacityCm3 { get; set; }
    public bool IsCargoLidOpen { get; set; }
    public string? DeviceId { get; set; }

    [Column(TypeName="geometry (point)")]
    public Point? Location { get; set; }

    public Guid? CompanyId { get; set; }
    public Company? Company { get; set; }
}

public enum RobotStatus
{
    Inactive,
    Idle,
    WaitingForCargo,
    Delivering,
    ReadyForPickup,
    Returning,
    Maintenance,
    Danger
}

File: DeliveryBot/deliverybot.client/src/interfaces/robot.ts
import { RobotStatus } from "./enums";
import { LocationDto } from "./geolocation";

export interface RobotInfoDto {
  name?: string | undefined;
  status?: RobotStatus;
  location?: LocationDto;
  batteryCharge?: number;
  weightCapacityG?: number;
  volumeCapacityCm3?: number;
  isCargoLidOpen?: boolean;
  deviceId?: string | undefined;
  companyId?: string | undefined;
  id?: string;
}

export interface GetDeliveryRobotQueryDto {
  name?: string | undefined;
  status?: RobotStatus;
  location?: LocationDto;
}

export interface CreateRobotCommand {
  name?: string | undefined;
  weightCapacityG?: number;
  volumeCapacityCm3?: number;
  deviceId?: string | undefined;
}

export interface EditRobotCommand {
  id?: string;
  name?: string | undefined;
  weightCapacityG?: number;
  volumeCapacityCm3?: number;
  deviceId?: string | undefined;
}

export interface SetRobotStatusCommand {
  id?: string;
  newStatus?: RobotStatus;
}

export interface ToggleRobotCargoLidCommand {
  robotId?: string;
}
